### PR TITLE
[ASCollectionNode] Add -isProcessingUpdates and -onDidFinishProcessingUpdates: APIs.

### PR DESCRIPTION
**Pull request by appleguy**
Over time, there have actually been a lot of legitimate uses for an API like this.
In fact, I'm not quite sure what has held us back from adding one!

I believe that at least some portion of -wait calls (even if less than 50%) could be
replaced with -onDidFinishProcessingUpdates:, which could potentially improve the
performance of applications using -wait by a significant amount.

Please take a close look at implementation correctness. Although I'm in a bit of a
rush, I'm aiming to make this properly documented and added a basic test -- but it
could certainly use some more detailed testing as a followup.

### CODE REVIEW

**Comment by appleguy** [comment]
This could be used to quite easily fix this ASBatchContext edge case: https://github.com/TextureGroup/Texture/issues/521

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/9da03b0b6493cc4128bfbec9bc0d81d0/raw/49ca1839495b42090dc5341de687f1cd3d9da9f5/log.txt)

**Comment by appleguy** [comment]
I see one of the new tests is failing on the build server, although it doesn't fail on my local machine - even trying 9.3, 10.0, and 10.3.1 simulators 

I'll wait for the last build to finish and then will start debugging-by-commit-pushing :).

File: Source/ASCollectionNode.h [hunk 1]
```diff
@@ -255,7 +255,36 @@ NS_ASSUME_NONNULL_BEGIN
...
+ *  ASCellNodes. If YES is returned, then calling -waitUntilAllUpdatesAreProcessed may take tens of
+ *  milliseconds to return as it blocks on these concurrent operations.
+ *
+ *  Returns NO if ASCollectionNode is fully synchronized with the underlying UICollectionView. This
+ *  means that until the next performBatchUpdates: is called, it is safe to compare UIKit values
+ *  (such as from UICollectionViewLayout) with your app's data source.
+ *
+ *  This method will always return NO if called immediately after -waitUntilAllUpdatesAreProcessed.
+ */
+- (BOOL)isProcessingUpdates;
```
Comment (Adlai-Holler) [inline Source/ASCollectionNode.h]: Trivial: Make this a readonly property for better swift interop?

File: Source/Details/ASDataController.mm [hunk 2]
```diff
@@ -437,6 +437,29 @@ - (void)waitUntilAllUpdatesAreCommitted
   [self _scheduleBlockOnMainSerialQueue:^{ }];
 }
 
+- (BOOL)isProcessingUpdates
+{
+  ASDisplayNodeAssertMainThread();
+  long waitResult = dispatch_group_wait(_editingTransactionGroup, DISPATCH_TIME_NOW);
+  return (waitResult == 0 ? NO : YES);
```
Comment (Adlai-Holler) [inline Source/Details/ASDataController.mm]: Is it possible that we could return a false `NO` if they call this in between the time the editing transaction group is finished and the main queue services the data controller's UIKit flush?

We could either flush the main serial queue (but not the editing transaction queue) or add API to check if the main serial queue is empty.

File: Source/Details/ASDataController.mm [hunk 3]
```diff
@@ -437,6 +437,29 @@ - (void)waitUntilAllUpdatesAreCommitted
...
+  } else {
+    dispatch_async(_editingTransactionQueue, ^{
+      // Retry the block. If we're done processing updates, it'll run immediately, otherwise
+      // wait again for updates to quiesce completely.
+      dispatch_async(dispatch_get_main_queue(), ^{
+        [self onDidFinishProcessingUpdates:completion];
+      });
+    });
+  }
+}
```
Comment (Adlai-Holler) [inline Source/Details/ASDataController.mm]: A couple low-impact notes on this method:

- If we're asserting main, is it worth removing `ASPerformBlockOnMainThread`? Presumably an error of this kind can be reliably caught during development. If the notion here is "better safe than sorry," I'm down with that as well.
- Should we schedule the retry on the main serial queue, so that if they wait on updates to complete we can be sure this block is run before returning from the wait? I suppose it's not a big deal – there's no guarantee this block will be run **as soon as** we finish processing updates. But at the same time, this is a break from the usual pattern of `editingQueue -> main serial queue` for main-thread-injection into data controller's async work.

File: Tests/ASCollectionViewTests.mm [hunk 4]
```diff
@@ -1050,8 +1050,17 @@ - (void)testInitialRangeBounds
...
+  XCTAssertTrue(cn.isProcessingUpdates, @"ASCollectionNode should still be processing updates after initial layoutIfNeeded call (reloadData)");
+  [cn onDidFinishProcessingUpdates:^{
+    XCTAssertTrue(!cn.isProcessingUpdates, @"ASCollectionNode should no longer be processing updates inside -onDidFinishProcessingUpdates: block");
+  }];
+
   // Wait for ASDK reload to finish
   [cn waitUntilAllUpdatesAreCommitted];
+
+  XCTAssertTrue(!cn.isProcessingUpdates, @"ASCollectionNode should no longer be processing updates after -wait call");
+
```
Comment (Adlai-Holler) [inline Tests/ASCollectionViewTests.mm]: 💕

File: Source/ASCollectionNode.h (hunk 1 above)
Comment (appleguy) [inline Source/ASCollectionNode.h]: Great suggestion, I'll try to remember this in the future (so foreign to me :-P)

File: Source/Details/ASDataController.mm (hunk 2 above)
Comment (appleguy) [inline Source/Details/ASDataController.mm]: Yeah, this is a good catch. I think this might be possible.

If we flush the main serial queue here, you could get situations where calling isProcessingUpdates results in calls out to the delegate / dataSource, which would be pretty weird and maybe unexpected by clients. So I think I'll just check the length of the queue.

Comment (appleguy) [inline Source/Details/ASDataController.mm]: @Adlai-Holler do you think we should perform a setNeedsLayout in the case that the queues are drained, before returning NO? This would technically be needed to ensure all information is flushed to / reflected in the UICollectionViewLayout...

File: Source/Details/ASDataController.mm (hunk 3 above)
Comment (appleguy) [inline Source/Details/ASDataController.mm]: Yes, the idea is some extra safety in the case the assertion fails in production. It is not likely to be necessary, but the semantics of PerformOnMain are clear enough that it seems OK.

Interested in your input on the other point. I had originally implemented it as waiting on the mainSerialQueue instead of a dispatch.

I was thinking (and now believe this may be incorrect) that doing it this way would allow us to call out to the blocks in an "intermediate quiesced" state to perform some operations before potentially more edits start getting processed.

However with the changes to isProcessingUpdates, I think this is actually invalid and we should schedule on the serial queue.

Comment (appleguy) [inline Source/Details/ASDataController.mm]: @Adlai-Holler is it correct that we do NOT perform using the _editingTransaction*Group*? I don't want multiple scheduled blocks to get stuck waiting for each other (as they do schedule in the main serial queue, and so I think this is correct, but I'm not completely confident.

File: Source/Details/ASMainSerialQueue.mm [hunk 5]
```diff
@@ -40,6 +40,11 @@ - (instancetype)init
   return self;
 }
 
+- (NSUInteger)numberOfScheduledBlocks
+{
+  return _blocks.count;
```
Comment (nguyenhuy) [inline Source/Details/ASMainSerialQueue.mm]: I think we need to grab `_serialQueueLock` lock here.

File: Source/Details/ASDataController.mm [hunk 6]
```diff
@@ -437,6 +437,35 @@ - (void)waitUntilAllUpdatesAreCommitted
   [self _scheduleBlockOnMainSerialQueue:^{ }];
 }
 
+- (BOOL)isProcessingUpdates
+{
+  ASDisplayNodeAssertMainThread();
+  if (_mainSerialQueue.numberOfScheduledBlocks > 0) {
+    return YES;
```
Comment (nguyenhuy) [inline Source/Details/ASDataController.mm]: 👍

File: Source/ASCollectionNode.h [hunk 7]
```diff
@@ -501,9 +530,13 @@ NS_ASSUME_NONNULL_BEGIN
...
  *
  * @deprecated This method is deprecated in 2.0. Use @c reloadDataWithCompletion: and
- *   then @c waitUntilAllUpdatesAreCommitted instead.
+ *   then @c waitUntilAllUpdatesAreProcessed instead.
  */
-- (void)reloadDataImmediately ASDISPLAYNODE_DEPRECATED_MSG("Use -reloadData / -reloadDataWithCompletion: followed by -waitUntilAllUpdatesAreCommitted instead.");
+- (void)reloadDataImmediately ASDISPLAYNODE_DEPRECATED_MSG("Use -reloadData / -reloadDataWithCompletion: followed by -waitUntilAllUpdatesAreProcessed instead.");
+
+// TODO: Rename framework uses of this method and add deprecation message.
+// ASDISPLAYNODE_DEPRECATED_MSG("This method has been renamed to -waitUntilAllUpdatesAreProcessed.");
```
Comment (nguyenhuy) [inline Source/ASCollectionNode.h]: If you're not going to do this in this PR, make sure to create an issue to track it.

**Comment by appleguy** [comment]
@nguyenhuy @Adlai-Holler thanks for the great feedback! I went ahead and renamed the method from Committed to Processing everywhere except the *View classes, where a deprecation warning already exists and now directs people to the Processing version (it seemed silly to add a second, also deprecated name).

This change makes me realize that we should give AS{Collection,Table}Node access to the ASDataController directly. There are a number of APIs where no participation from the View class is required. Even without breaking support for using the View classes independently (which is becoming less and less important), calls like isProcessingUpdates or waitUntil.. could be passed directly from the Node to ASDataController.

**Comment by nguyenhuy** [comment]
@appleguy: Yes, agree that the nodes should have direct access to data controller. In a far future, we may even tell the data controller to do the initial load before the view is even allocated. Definitely possible with an async collection layout.

File: Source/Details/ASDataController.mm (hunk 2 above)
Comment (Adlai-Holler) [inline Source/Details/ASDataController.mm]: Are you maybe thinking about `layoutIfNeeded`? That would absolutely make sense 👍

File: Source/Details/ASDataController.mm (hunk 3 above)
Comment (Adlai-Holler) [inline Source/Details/ASDataController.mm]: That's right! The transaction group is only (and always) applied to work blocks on the editing transaction queue.

**Comment by nguyenhuy** [comment]
ok, let's get this in!

**Comment by appleguy** [comment]
Thanks Huy!
